docs: part 2 of compactor documents with best practice and guidelines (#5880)
influxdata/influxdb_iox: docs/compactor.md

   - Level-0 files are small files ingested by the Ingesters
   - Level-1 files are files created by a Compactor as a result of compacting one or many level-0 files with their overlapped level-1 files.
   - Level-2 files are files created by a Compactor as a result of compacting one or many level-1 files with their overlapped level-2 files.
  

> Regarless of level, a file in IOx must belong to a partition which represents data of a time range which is usually a day. Two files of different partitions never overlap in time and hence the Compactor only needs to compact files that belong to the same partition.

Reviewer: `Regarless` should be `Regardless`

# Memory Estimation

The idea of a single compaction is to compact as many small input files as possible into one or few larger output files as follows

```text
     ┌────────┐         ┌────────┐     
     │        │         │        │     
     │ Output │  .....  │ Output │     
     │ File 1 │         │ File m │     
     │        │         │        │     
     └────────┘         └────────┘     
          ▲                  ▲         
          │                  │         
          │                  │         
      .─────────────────────────.      
 _.──'                           `───. 
(               Compact               )
 `────.                         _.───' 
       `───────────────────────'       
       ▲        ▲              ▲       
       │        │              │       
       │        │              │       
   ┌───────┐┌──────┐       ┌──────┐    
   │       ││      │       │      │    
   │ Input ││Input │       │Input │    
   │File 1 ││File 2│ ..... │File n│    
   │       ││      │       │      │    
   └───────┘└──────┘       └──────┘    
Figure 1: Compact a Partition
```

> Currently, in order to run that plan, we need to compute memory needed for loading full input files into memory, memory for streaming input files in parallel, and memory for output streams. The boxes in the diagram below illustrates the memory needed to run the query plan above. IOx picks number of input files to compact based on their sizes, number of columns and columns types of the file's table, max desired output files, and the memory budget the compactor is provided. Details:

Reviewer 2: ```suggestion
Currently, in order to avoid over committing memory (and OOMing), the compactor computes an estimate of the memory needed for loading full input files into memory, memory for streaming input files in parallel, and memory for output streams. The boxes in the diagram below illustrates the memory needed to run the query plan above. IOx picks number of input files to compact based on their sizes, number of columns and columns types of the file's table, max desired output files, and the memory budget the compactor is provided. Details:
```

          │           │          │            │     
          └───────────┘          └────────────┘     
                                                    
               ▲                        ▲           
               │                        │           
               │                        │           
               │                        │           
                 .─────────────────────.            
          _.────'                       `─────.     
       ,─'                                     '─.  
      ╱            Run Compaction Plan            ╲ 
     (                                             )
      `.                                         ,' 
        '─.                                   ,─'   
           `─────.                     _.────'      
                  `───────────────────'             
       ▲              ▲                      ▲      
       │              │                      │      
       │              │                      │      
       │              │                      │      
┌────────────┐ ┌────────────┐         ┌────────────┐
│ Memory for │ │ Memory for │         │ Memory for │

> │ Streamming │ │ Streamming │         │ Streamming │

Reviewer: streaming has just one `m`

└────────────┘ └────────────┘         └────────────┘
┌────────────┐ ┌────────────┐         ┌────────────┐
│ Memory for │ │ Memory for │         │ Memory for │
│Loading Full│ │Loading Full│  .....  │Loading Full│
│   File 1   │ │   File 2   │         │   File n   │
│            │ │            │         │            │
└────────────┘ └────────────┘         └────────────┘
Figure 2:  Memory Estimation for a Compaction Plan
```

- Memory for loading a full file : Twice the file size.
- Memory for streaming an input file: See [estimate_arrow_bytes_for_file](https://github.com/influxdata/influxdb_iox/blob/bb7df22aa1783e040ea165153876f1fe36838d4e/compactor/src/compact.rs#L504) for the details but, briefly, each column of the file will need `size_of_a_row_of_a_column * INFLUXDB_IOX_COMPACTION_MIN_ROWS_PER_RECORD_BATCH_TO_PLAN` in which `size_of_a_row_of_a_column` depends on column type and the file cardinality.
- Memory for an output stream is similar to memory for streaming an input file. Number of output streams is estimated based on the sizes of the input files and max desired file size `INFLUXDB_IOX_COMPACTION_MAX_DESIRED_FILE_SIZE_BYTES`.
 

> IOx limits number of input files using the budget it is provided in `INFLUXDB_IOX_COMPACTION_MEMORY_BUDGET_BYTES`. However, IOx also caps number of input files based on `INFLUXDB_IOX_COMPACTION_MAX_COMPACTING_FILES` even if it has a lot of memory budget.

Reviewer 2: ```suggestion
IOx limits the number of input files using the budget provided in `INFLUXDB_IOX_COMPACTION_MEMORY_BUDGET_BYTES`. However, IOx also caps number of input files based on `INFLUXDB_IOX_COMPACTION_MAX_COMPACTING_FILES` even if more could fit under the memory budget.
```

**Best Pratice Recommendation for your configuration:**
- INFLUXDB_IOX_COMPACTION_MEMORY_BUDGET_BYTES: 1/3 (or at most 1/2) your total actual memory.
- INFLUXDB_IOX_COMPACTION_MAX_COMPACTING_FILES: 20
- INFLUXDB_IOX_COMPACTION_MIN_ROWS_PER_RECORD_BATCH_TO_PLAN: 32 * 1024
- INFLUXDB_IOX_COMPACTION_MAX_DESIRED_FILE_SIZE_BYTES: 100 * 1024 * 1024
- INFLUXDB_IOX_COMPACTION_PERCENTAGE_MAX_FILE_SIZE: 5

# Avoid and deal with files in skipped_compactions
To deduplicate data correctly, the Compactor must compact level-0 files in ascending order of their sequence nubers and with their overlapped level-1 files. If the first level-0 and its overlapped level-1 files are too large and their memory estimation in Figure 2 is over the budget defined in INFLUXDB_IOX_COMPACTION_MEMORY_BUDGET_BYTES, the compactor won't be able to compact that partition. To avoid considering that same partition again and again, the compactor will put that partition into the catalog table `skipped_compactions`.

> If you find your queries on data of a partition that is in `skipped_compactions` are slow, we may want to force the Compactor to compact that partition by increasing your memory budget and then removing that partition from the `skipped_compactions`. If you remove the partition without adjusting your config params, the Compactor will put its back in again without compacting it.

Reviewer: does one remove it from the skipped table manually with an sql statement straight on the DB? might be worth giving an example

Reviewer 2: If we find ourselves or others needing to do this semi regularly I think we should invest in a compactor gRPC API rather than modifying the production catalog

Author: Good idea

Author: @lukebond and @alamb 
I have added this: 
"We are working on an gRPC API to let you see the content of the `skipped_compactions` and remove a specific partition from it. In the mean time, you can do this by using delete SQL statement directly from your Catalog (see SQL section for this  statement) but we do not recommend modifying the Catalog unless you know your partitions and their data files very well."

I will open a ticket to implement them

Author: Ticket https://github.com/influxdata/influxdb_iox/issues/5894

> Increasing memory budget can be as simply as increasing your actual memory size and then increasing the value of INFLUXDB_IOX_COMPACTION_MEMORY_BUDGET_BYTES accordingly. You can also try to reduce the value of INFLUXDB_IOX_COMPACTION_MIN_ROWS_PER_RECORD_BATCH_TO_PLAN to minimum 8 * 1024 but you may hit OOMs if you do so. This depends on column types and number of columns of your file.

Reviewer: `can be as simple as`

If your partition is put into the  `skipped_partition` with the reason `over limit of num files`, you have to increase INFLUXDB_IOX_COMPACTION_MAX_COMPACTING_FILES accordingly but you may hit OOMs if you do not increase your actual memory.

# Avoid Deduplication in Querier

> Deduplication is known to be expensive. To avoid deduplication work during query time in Queriers, your files should not be overalpped in time range. This can be achived by having all files of a partition in either level-1 or level-2. With current design, if your compactor catches up well, partitions with recent level-0 files within the last 4 hours should have at most 2 level-2 files. Partition without new level-0 files the last 8 hours should have all level-2 files. Depedning on the performance in Querier, we can adjust the Compactor (a future feature) to have all files in level-1 or level-2.

Reviewer: `overlapped` typo.
i think `achived` should be `achieved`
`With the current design`
`Partitions without new level-0 files in the last 8 hours`
`Depending on the performance in the Querier`
